File: src/nasdaq/DecodeFastLogExtended.java
//// compile in Bash: javac DecodeFastLogExtended.java
//// then run in Bash: java DecodeFastLogExtended
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Date;

public class DecodeFastLogExtended {
    static final String[] MESSAGE_TYPES = {
        "UNKNOWN",
        "ORDER_ADD",
        "ORDER_EXEC",
        "ORDER_CANCEL",
        "DEPTH_UPDATE",
        "MARKET_STATUS",
        "SPECIAL_NEWS"
    };

    public static void decodeLogFile(String inputFile, String outputCsv) throws Exception {
        try (
            DataInputStream in = new DataInputStream(new BufferedInputStream(new FileInputStream(inputFile)));
            PrintWriter out = new PrintWriter(new OutputStreamWriter(new FileOutputStream(outputCsv), StandardCharsets.UTF_8))
        ) {
            out.println("Timestamp,MessageType,OrderID,Volume,Price,ExecQty,DepthLevel,Bid,Ask/Status/Headline");

            while (in.available() > 0) {
                byte msgType = in.readByte();
                double timestampVal = in.readDouble();

                long seconds = (long) timestampVal;
                long millis = (long) ((timestampVal - seconds) * 1000);
                Date date = new Date(seconds * 1000 + millis);
                String timestampStr = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS").format(date);

                String msgTypeStr = msgType >= 0 && msgType < MESSAGE_TYPES.length ? MESSAGE_TYPES[msgType] : "UNKNOWN";

                if (msgType == 1) { // ORDER_ADD
                    long orderId = Integer.toUnsignedLong(in.readInt());
                    long volume = Integer.toUnsignedLong(in.readInt());
                    float price = in.readFloat();
                    System.out.printf("%s | ORDER_ADD: ID=%d, Vol=%d, Price=%.2f%n", timestampStr, orderId, volume, price);
                    out.printf("%s,%s,%d,%d,%.2f,,,,%n", timestampStr, msgTypeStr, orderId, volume, price);
                } else if (msgType == 2) { // ORDER_EXEC
                    long orderId = Integer.toUnsignedLong(in.readInt());
                    long execQty = Integer.toUnsignedLong(in.readInt());
                    System.out.printf("%s | ORDER_EXEC: ID=%d, Qty=%d%n", timestampStr, orderId, execQty);
                    out.printf("%s,%s,%d,,,%d,,,,%n", timestampStr, msgTypeStr, orderId, execQty);
                } else if (msgType == 3) { // ORDER_CANCEL
                    long orderId = Integer.toUnsignedLong(in.readInt());
                    System.out.printf("%s | ORDER_CANCEL: ID=%d%n", timestampStr, orderId);
                    out.printf("%s,%s,%d,,,,,,,%n", timestampStr, msgTypeStr, orderId);
                } else if (msgType == 4) { // DEPTH_UPDATE
                    long level = Integer.toUnsignedLong(in.readInt());
                    float bid = in.readFloat();
                    float ask = in.readFloat();
                    System.out.printf("%s | DEPTH_UPDATE: Level=%d, Bid=%.2f, Ask=%.2f%n", timestampStr, level, bid, ask);
                    out.printf("%s,%s,,,,%d,%.2f,%.2f,%n", timestampStr, msgTypeStr, level, bid, ask);
                } else if (msgType == 5) { // MARKET_STATUS
                    byte status = in.readByte();
                    String statusStr = status == 0 ? "CLOSED" : status == 1 ? "OPEN" : status == 2 ? "HALT" : "UNKNOWN";
                    System.out.printf("%s | MARKET_STATUS: %s%n", timestampStr, statusStr);
                    out.printf("%s,%s,,,,,,,%s%n", timestampStr, msgTypeStr, statusStr);
                } else if (msgType == 6) { // SPECIAL_NEWS
                    byte[] headlineBytes = new byte[50];
                    in.readFully(headlineBytes);
                    int headlineLen = 0;
                    for (; headlineLen < headlineBytes.length && headlineBytes[headlineLen] != 0; headlineLen++);
                    String headline = new String(headlineBytes, 0, headlineLen, StandardCharsets.UTF_8);
                    System.out.printf("%s | SPECIAL_NEWS: %s%n", timestampStr, headline);
                    out.printf("%s,%s,,,,,,,%s%n", timestampStr, msgTypeStr, headline.replace(",", " "));
                } else {
                    // Unknown type, break
                    break;
                }
            }
        }
        System.out.println("\n[✓] Decoded messages saved to decoded_fast_log.csv");
    }

    public static void main(String[] args) throws Exception {
        String inputFile = "fast_extended_10k.log";
        String outputCsv = "decoded_fast_log.csv";
        decodeLogFile(inputFile, outputCsv);
    }
}

File: src/nasdaq/DecodeFastLogExtended.kt
/// compile in bash: kotlinc DecodeFastLogExtended.kt -include-runtime -d DecodeFastLogExtended.jar
/// run in bash: java -jar DecodeFastLogExtended.jar
import java.io.*
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.Charset
import java.text.SimpleDateFormat
import java.util.*

val MESSAGE_TYPES = arrayOf(
    "UNKNOWN",
    "ORDER_ADD",
    "ORDER_EXEC",
    "ORDER_CANCEL",
    "DEPTH_UPDATE",
    "MARKET_STATUS",
    "SPECIAL_NEWS"
)

fun main() {
    val inputFile = "fast_extended_10k.log"
    val outputCsv = "decoded_fast_log.csv"
    decodeLogFile(inputFile, outputCsv)
}

fun decodeLogFile(inputFile: String, outputCsv: String) {
    DataInputStream(BufferedInputStream(FileInputStream(inputFile))).use { infile ->
        PrintWriter(OutputStreamWriter(FileOutputStream(outputCsv), Charset.forName("UTF-8"))).use { outfile ->
            outfile.println("Timestamp,MessageType,OrderID,Volume,Price,ExecQty,DepthLevel,Bid,Ask/Status/Headline")

            while (infile.available() > 0) {
                val msgType = infile.readUnsignedByte()
                val timestampVal = infile.readDouble()
                val seconds = timestampVal.toLong()
                val millis = ((timestampVal - seconds) * 1000).toLong()
                val date = Date(seconds * 1000 + millis)
                val timestampStr = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS").format(date)
                val msgTypeStr = if (msgType in MESSAGE_TYPES.indices) MESSAGE_TYPES[msgType] else "UNKNOWN"

                when (msgType) {
                    1 -> { // ORDER_ADD
                        val orderId = infile.readInt().toUInt()
                        val volume = infile.readInt().toUInt()
                        val price = infile.readFloat()
                        println("$timestampStr | ORDER_ADD: ID=$orderId, Vol=$volume, Price=$price")
                        outfile.println("$timestampStr,$msgTypeStr,$orderId,$volume,$price,,,,")
                    }
                    2 -> { // ORDER_EXEC
                        val orderId = infile.readInt().toUInt()
                        val execQty = infile.readInt().toUInt()
                        println("$timestampStr | ORDER_EXEC: ID=$orderId, Qty=$execQty")
                        outfile.println("$timestampStr,$msgTypeStr,$orderId,,,$execQty,,,,")
                    }
                    3 -> { // ORDER_CANCEL
                        val orderId = infile.readInt().toUInt()
                        println("$timestampStr | ORDER_CANCEL: ID=$orderId")
                        outfile.println("$timestampStr,$msgTypeStr,$orderId,,,,,,,")
                    }
                    4 -> { // DEPTH_UPDATE
                        val level = infile.readInt().toUInt()
                        val bid = infile.readFloat()
                        val ask = infile.readFloat()
                        println("$timestampStr | DEPTH_UPDATE: Level=$level, Bid=$bid, Ask=$ask")
                        outfile.println("$timestampStr,$msgTypeStr,,,,$level,$bid,$ask,")
                    }
                    5 -> { // MARKET_STATUS
                        val status = infile.readUnsignedByte()
                        val statusStr = when (status) {
                            0 -> "CLOSED"
                            1 -> "OPEN"
                            2 -> "HALT"
                            else -> "UNKNOWN"
                        }
                        println("$timestampStr | MARKET_STATUS: $statusStr")
                        outfile.println("$timestampStr,$msgTypeStr,,,,,,,$statusStr")
                    }
                    6 -> { // SPECIAL_NEWS
                        val headlineBytes = ByteArray(50)
                        infile.readFully(headlineBytes)
                        val headlineLen = headlineBytes.indexOf(0).let { if (it < 0) headlineBytes.size else it }
                        val headline = String(headlineBytes, 0, headlineLen, Charset.forName("UTF-8"))
                        println("$timestampStr | SPECIAL_NEWS: $headline")
                        outfile.println("$timestampStr,$msgTypeStr,,,,,,,$headline")
                    }
                    else -> break
                }
            }
        }
    }
    println("\n[✓] Decoded messages saved to $outputCsv")
}
